### ISSUE TITLE
OSX cross compiling to any target is broken when plugins are used

### ISSUE THREAD

**Issue by metajack** [rust-lang/rust#21703]
@kmcallister reported this when doing the Rust alpha upgrade for Servo.

**Comment by metajack** [comment]
cc @alexcrichton

**Comment by alexcrichton** [comment]
Hm I can't seem to reproduce this any more. One thing I noticed is to be sure that plugins are tagged with `#[no_link]`. Do you have a repo I could reproduce this on?

**Comment by kmcallister** [comment]
This report is via @larsbergstrom; I don't have all the details.

**Comment by larsbergstrom** [comment]
@alexcrichton You can repro this in our codebase by building for Android on an OSX box (https://github.com/servo/servo/wiki/Building-for-Android ).

You will get:
```
   Compiling android_glue v0.0.1 (file:///Users/larsberg/servo/support/android-rs-glue)
error: extern location for compile_msg is of an unknown type: /Users/larsberg/servo/components/servo/target/deps/libcompile_msg-03046bbb505a03fb.dylib
error: extern location for compile_msg is of an unknown type: /Users/larsberg/servo/components/servo/target/deps/libcompile_msg-03046bbb505a03fb.dylib
/Users/larsberg/servo/support/android-rs-glue/glue/src/lib.rs:8:1: 8:26 error: can't find crate for `compile_msg`
/Users/larsberg/servo/support/android-rs-glue/glue/src/lib.rs:8 extern crate compile_msg;
                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to 3 previous errors
```

**Comment by alexcrichton** [comment]
Hm, I don't know why, but it looks like this may be fixed on master. I compiled this with today's nightly:

```rust
#![no_std]
#![feature(plugin)]
#[plugin]
#[no_link]
extern crate compile_msg;
```

and everything worked. I downloaded master servo and compiled with the version of the compiler you're pinned to and it got the same error message.

I'm not quite sure what would have fixed this, but I guess you guys should expect a fix in the next rust upgrade?

**Comment by larsbergstrom** [comment]
OK, I will try it on 52c74e63dacd49017b19330e0cbecbac0a3fe62e and see how that goes.

I assume you did build a cross-toolchain and compile for Android (i.e., "./mach build --android")? This does not break on OSX targeting OSX. I only ask because nightlies don't have cross-toolchains.

**Comment by alexcrichton** [comment]
I never quite got all the way to getting Servo itself built, I ended up working to minimize it to a much smaller instance (hence my wariness to close this issue).

I think the bits that I minimized down to would have the same behavior as Servo itself, but I'm not 100% sure :(

**Comment by larsbergstrom** [comment]
Mostly fixed with newer rustc - please close.

**Comment by larsbergstrom** [comment]
Gah, this is still not fixed.

**Comment by steveklabnik** [comment]
Is this still true today?

**Comment by larsbergstrom** [comment]
Nope, feel free to close now.

**Comment by alexcrichton** [comment]
Ok, thanks @larsbergstrom!
